File: lib/rallhook.rb
=begin

This file is part of the rallhook project, http://github.com/tario/

Copyright (c) 2009-2010 Roberto Dario Seminara <robertodarioseminara@gmail.com>

rallhook is free software: you can redistribute it and/or modify
it under the terms of the gnu general public license as published by
the free software foundation, either version 3 of the license, or
(at your option) any later version.

rallhook is distributed in the hope that it will be useful,
but without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.  see the
gnu general public license for more details.

you should have received a copy of the gnu general public license
along with rallhook.  if not, see <http://www.gnu.org/licenses/>.

=end
require "rubygems"
require "rallhook_base"

module RallHook
#
#Internal class used for return method redirection messages
#Example:
# ...
# class X
#   def foo
#   end
# end
# def handle_method
#   return RallHook::Redirect.new(X, X.new, :foo)
# end
# ...
#
#Note: Use x.redirect(:foo) instead (see Object#redirect )
#
  class Redirect
		include MethodRedirect

		def initialize(klass, recv, m, unhook = nil)
			@klass = klass
			@recv = recv
			@method = m
      @unhook = unhook
		end
	end
#
#Internal class used by RallHook::Helper#return_value
#
	class ReturnHolder
	 	def initialize(v)
			@value = v
    end
    def value(*x)
      @value
    end
  end

#
#This module brings together classes and methods to facilitate the handling
#
	module Helper

#Equivalent to:
# recv.redirect(m, klass)
		def redirect_call(klass, recv, m)
			::RallHook::Redirect.new(klass,recv,m)
		end

#Return a value as return value of the method being handled
#Example:
#
# include RallHook::Helper
# def handle_method
#   return_value(4) # all methods called returns 4
# end
#
		def return_value(v)
      recv = ReturnHolder.new(v)
      klass = recv.class
      m = :value
			redirect_call(klass, recv, m)
	  end

#
#This class acts as base to define classes that acts as receptors and wrappers of
#redirected methods. Allows interception of methods with its parameters and controlled recall
#in a hooking logic
#
#Example 1: basic modifications of arguments
#
# class MethodHandler
#  include RallHook::Helper
#
#  class FooMethodWrapper < RallHook::MethodWrapper
#    def call(foo_argument)
#      original_call(foo_argument + 5) # add 5 in all calls to foo
#    end
#  end
#
#  def handle_method (klass,recv,m,method_id)
#   if m == :foo
#    FooMethodWrapper.redirect_handler(klass,recv,m,method_id)
#   else
#    nil
#   end
#  end
# end
#
# ... hooking using MethodHandler ( see Hook#hook, README and examples)
#
#
#Example 2: generic recall of methods
#
# class MethodHandler
#  include RallHook::Helper
#
#  class GenericMethodWrapper < RallHook::MethodWrapper
#    def call(*x)
#      # call with reyield if block_given
#      if block_given?
#        original_call(*x) do |*a|
#          yield(*a)
#        end
#      else
#        original_call(*x)
#      end
#    end
#  end
#
#  def handle_method (klass,recv,m,method_id)
#    return GenericMethodWrapper.redirect_handler(klass,recv,m,method_id)
#  end
# end
#
# ... hooking using MethodHandler ( see Hook#hook, README and examples)
#
#
    class MethodWrapper

      attr_accessor :klass, :recv, :method_name, :method_id

#Reactivate the hook status to intercept methods
#Allow to pass a block to enable the hook status only in that block
#
#Example:
# class MethodHandler
#  include RallHook::Helper
#
#  class FooMethodWrapper < RallHook::MethodWrapper
#    def call(foo_argument)
#      original_call(foo_argument + 5) # add 5 in all calls to foo
#      rehook do  # the print "hello world hooked" are intercepted too
#         print "hello world hooked\n"
#      end
#    end
#  end
#
#  ... definition of handle_method that use  FooMethodWrapper as redirect, etc...
#
#Note: is not necesary to use rehook to hook the nested calls in original_call
#MethodWrapper#original_call does that internally
#
      def rehook
        if block_given?
          ::RallHook::Hook.rehook do
            yield
          end
        else
          ::RallHook::Hook.rehook
        end
      end

#
#Same as Hook#from
#
      def from(a)
        ::RallHook::Hook.from(a)
        self
      end

#
#Recall the original method(specified in initialization parameters)
#
      def original_call(*args)
       mname = self.method_name
       mklass = self.klass
       mid = self.method_id
       recv_ = self.recv

       if block_given?
         from(4).rehook do
           recv_.method(mklass,mid).call(*args) do |*blockargs|
             yield(*blockargs)
           end
         end
       else
         from(4).rehook do
           recv_.method(mklass,mid).call(*args)
         end
       end
      end

#
#Target of redirection, should be implemented to intercept the call
#
      def call(*args)
      end

      def call_with_rehook(*args)
        if block_given?
          call(*args) do |*x|
            yield(*x)
          end
        else
          call(*args)
        end
      ensure
        rehook
      end

#
#Makes a redirect_handler to wrap a method call with this MethodWrapper. Example:
#
# FooMethodWrapper.redirect_handler(klass,recv,m,method_id)
#
      def self.redirect_handler(klass,recv,method_name, method_id)
        mw = self.new
        mw.klass = klass
        mw.recv = recv
        mw.method_name = method_name
        mw.method_id = method_id
        mw.redirect_with_unhook(:call_with_rehook)
      end

    end
  end



class HookHandler

#default method_handler of a Hook does nothing with any method call
  def handle_method
    nil
  end

#enable the hook. Example
#
#  class MethodHandler < Hook
#    def handle_method
#      nil # do nothing
#    end
#  end
#
#  MethodHandler.hook do
#     print "hello world\n"
#  end
#

  def hook
    if block_given?
      ::RallHook::Hook.hook self do
        yield
      end
    else
      ::RallHook::Hook.hook self
    end
  end

#Instance handler and activate the hook
#Example:
#
#  class MethodHandler < Hook
#    def handle_method
#      nil # do nothing
#    end
#  end
#
#  MethodHandler.hook do
#    print "hello world\n"
#  end

  def self.hook
    if block_given?
      self.new.hook do
        yield
      end
    else
      self.new.hook
    end
  end
end



end

# To facilitate the use of hooking/redirection features,
# rallhook defines some methods in the Object class
#

class Object


# Same as redirect, but disable the hook/interception of methods (dont call handle_redirect) before
# make the call, the hook status may be restored with Hook#rehook
# Anyway, it is desirable to use MethodWrapper instead to "wrap" method calls
#
# see RallHook::Helper::MethodWrapper
#
  def redirect_with_unhook(method_name, klass = nil)
    if klass
      ::RallHook::Redirect.new(klass,self,method_name,true)
    else
      ::RallHook::Redirect.new(self.class,self,method_name,true)
    end

  end

# Generates a redirect message to return it in handle_method to cause the redirection of
# the method that is beign processed with the object as receiver, the method_name the new target
# and the class as second parameter if specified
#
# Example:
#   class X
#     def foo(*args)
#     end
#   end
#
#   class MethodHandler
#     @@x = x.new
#     def handle_method(klass, recv, m , method_id)
#       if m == :bar then
#         # redirect bar calls to foo in x
#         return x.redirect(:foo)
#       end
#       nil # do nothing
#     end
#   end
#
#   # hook using MethodHandler, etc... (see README and examples)
#
#
  def redirect(method_name, klass = nil)
		if klass
			::RallHook::Redirect.new(klass,self,method_name)
		else
			::RallHook::Redirect.new(self.class,self,method_name)
		end
	end
end

File: ext/rallhook_base/rallhook.c
/*

This file is part of the rallhook project, http://github.com/tario/rallhook

Copyright (c) 2009-2010 Roberto Dario Seminara <robertodarioseminara@gmail.com>

rallhook is free software: you can redistribute it and/or modify
it under the terms of the gnu general public license as published by
the free software foundation, either version 3 of the license, or
(at your option) any later version.

rallhook is distributed in the hope that it will be useful,
but without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.  see the
gnu general public license for more details.

you should have received a copy of the gnu general public license
along with rallhook.  if not, see <http://www.gnu.org/licenses/>.

*/

#include <ruby.h>
#include "method_node.h"
#include "ruby_redirect.h"
#include "restrict_def.h"

VALUE rb_cHook;
VALUE rb_mRallHook;
VALUE rb_mMethodRedirect;
VALUE rb_mMethodReturn;
ID id_call_;

ID id_call;
ID id_method_wrapper;
ID id_handle_method;
ID id_binding;
ID id_method_added;
ID id_hook_enabled;
ID id_hook_enable_left;
ID id_hook_proc;

ID id_return_value_var, id_klass_var, id_recv_var, id_method_var, id_unhook_var;

void *rb_get_method_body(VALUE klass, ID id, ID *idp);

typedef struct rb_thread_struct
{
    VALUE self;
	void *vm; // rb_vm_t_

    /* execution information */
    VALUE *stack;		/* must free, must mark */
    unsigned long stack_size;
    void *cfp; // rb_control_frame_t_
    int safe_level;
    int raised_flag;
    VALUE last_status; /* $? */

    /* passing state */
    int state;

    /* for rb_iterate */
    void *passed_block; // rb_block_t_

    // ...
} rb_thread_t__;

typedef struct AttachedThreadInfo_ {
	int hook_enabled;
	int hook_enable_left;
	VALUE hook_proc;
	int handle_method_arity;
} AttachedThreadInfo;

AttachedThreadInfo* tinfo_from_thread(VALUE thread) {
	VALUE tmp = rb_ivar_get( thread, rb_intern("__tinfo") );

	if (tmp == Qnil) {
		AttachedThreadInfo* tinfo = malloc(sizeof(AttachedThreadInfo));
		tinfo->hook_enabled = 0;
		tinfo->hook_enable_left = 0;
		tinfo->hook_proc = Qnil;

		VALUE tinfo_obj = Data_Wrap_Struct(rb_cObject, 0, 0,tinfo);

		rb_ivar_set( thread, rb_intern("__tinfo"), tinfo_obj);

		return tinfo;
	} else {

		AttachedThreadInfo* tinfo;
		Data_Get_Struct(tmp, struct AttachedThreadInfo, tinfo);

		return tinfo;
	}
}

void redirect_left(AttachedThreadInfo* tinfo, int left) {
	tinfo->hook_enable_left = left;
}

void enable_redirect(AttachedThreadInfo* tinfo) {
	tinfo->hook_enabled = 1;
}

void disable_redirect(AttachedThreadInfo* tinfo) {
	tinfo->hook_enabled = 0;
}

int get_hook_enable_left(AttachedThreadInfo* tinfo) {
	return tinfo->hook_enable_left;
}

int get_hook_enabled(AttachedThreadInfo* tinfo) {
	return tinfo->hook_enabled;
}

VALUE get_hook_proc() {
	return tinfo_from_thread( rb_thread_current() )->hook_proc ;
}

/*
Disable the hook. Is not usually necesary because of the RAII feature of Hook#hook
*/
VALUE unhook(VALUE self) {
	disable_redirect(tinfo_from_thread( rb_thread_current() ) );
	return Qnil;
}

VALUE ensured_handle_method( VALUE params ) {
	int argc = (int)((VALUE*)params)[0];
	VALUE* argv = ((VALUE*)params)+1;
	return rb_funcall2( get_hook_proc(), id_handle_method, argc, argv);
}

VALUE restore_hook_status(AttachedThreadInfo* tinfo) {
	enable_redirect(tinfo);
	return Qnil;
}

void set_handle_method_arity(int value) {
	tinfo_from_thread(rb_thread_current() )->handle_method_arity = value;
}

void rallhook_redirect_handler ( VALUE* klass, VALUE* recv, ID* mid ) {

	VALUE current_thread = rb_thread_current();
	AttachedThreadInfo* tinfo = tinfo_from_thread(current_thread);
	int hook_enable_left  = get_hook_enable_left(tinfo);

	if (get_hook_enabled(tinfo) == 0 || hook_enable_left > 0){
		if(hook_enable_left>0) hook_enable_left --;
		redirect_left( tinfo, hook_enable_left );
		return;
	}

	VALUE argv_[6];
	if(*mid == id_method_added) {
		*recv = unshadow(*recv); // recv is THE class
		*klass = CLASS_OF(*recv);
	}

	argv_[0] = tinfo->handle_method_arity;
	if (tinfo->handle_method_arity == 4) {

		VALUE sym;
	    if (rb_id2name(*mid)) {
	    	sym = ID2SYM(*mid);
	    } else {
	    	sym =Qnil;
	    }


		argv_[1] = unshadow(*klass);
		argv_[2] = *recv;
		argv_[3] = sym;
		argv_[4] = LONG2FIX(*mid);
	} else {
		argv_[1] = unshadow(*klass);
		argv_[2] = *recv;
		argv_[3] = LONG2FIX(*mid);
	}

	ID original_id = *mid;

	disable_redirect(tinfo);

	rb_thread_t__* th;
	Data_Get_Struct( current_thread, rb_thread_t__, th );

	void* blockptr = th->passed_block;
	VALUE result = rb_ensure(ensured_handle_method,(VALUE)argv_,restore_hook_status, (VALUE)tinfo);

	th->passed_block = blockptr;

	if (original_id != id_binding ) {

		// method named "binding" cannot be redirected
		if (rb_obj_is_kind_of(result,rb_mMethodRedirect) == Qtrue ) {
			*klass = rb_ivar_get(result,id_klass_var );
			*recv = rb_ivar_get(result,id_recv_var );
			*mid = rb_to_id( rb_ivar_get(result,id_method_var) );

			if (rb_ivar_get(result,id_unhook_var) != Qnil ) {
				disable_redirect(tinfo);
			}

		} else {
			shadow_redirect(klass, recv, mid);
		}
	}

	// methods over class hook are illegal, may change the state of hook
	if (*recv == rb_cHook ) {
		rb_raise(rb_eSecurityError, "Illegal method call: Hook.%s", rb_id2name(*mid) );
	}


}

/*
Activate the hook, it is desirable to use the RAII call to make the hook block exception safe:

	rallhook.hook method_hadler do
		print "hello world\n" # calls to print, write or whatever are intercepted by method_handler#method_handle
	end # in the finish of the block, the hook are disabled

*/
VALUE hook(VALUE self, VALUE hook_proc) {

	tinfo_from_thread( rb_thread_current() )->hook_proc = hook_proc;

	VALUE handle_method_method =rb_obj_method(hook_proc, ID2SYM(id_handle_method) );
	VALUE handle_method_method_arity = rb_funcall( handle_method_method, rb_intern("arity"), 0 );
	tinfo_from_thread( rb_thread_current() )->handle_method_arity = FIX2INT( handle_method_method_arity );

	put_redirect_handler( rallhook_redirect_handler );

	enable_redirect(tinfo_from_thread(rb_thread_current()));

	if (rb_block_given_p() ) {
		return rb_ensure(rb_yield, Qnil, unhook, self);
	}

	return Qnil;
}


/*
Disable the hook in the next N calls and reenable them. Useful to avoid infinite recursion and used
in RallHook::Helper::MethodWrapper
*/
VALUE from(VALUE self, VALUE num) {
	redirect_left(tinfo_from_thread( rb_thread_current() ), FIX2INT(num)+1);
	return self;
}


/*
Re-enable the hook if the hook was disabled and was activated previously with Hook#hook
If no call to Hook#hook has made, rehook does nothing
*/
VALUE rehook(VALUE unused) {
	enable_redirect(tinfo_from_thread( rb_thread_current() ));
	if (rb_block_given_p() ) {
		return rb_ensure(rb_yield, Qnil, unhook, Qnil);
	}
	return Qnil;
}



VALUE thread_restore_attributes(AttachedThreadInfo* tinfo) {

	AttachedThreadInfo* current_tinfo = tinfo_from_thread(rb_thread_current());

	current_tinfo->hook_enabled = tinfo->hook_enabled;
	current_tinfo->hook_enable_left = tinfo->hook_enable_left;
	current_tinfo->hook_proc = tinfo->hook_proc;
	current_tinfo->handle_method_arity = tinfo->handle_method_arity;

	return Qnil;
}

VALUE rb_thread_acquire_attributes( VALUE thread ) {

	AttachedThreadInfo* orig = tinfo_from_thread(thread);
	AttachedThreadInfo* dest = tinfo_from_thread(rb_thread_current() );

	AttachedThreadInfo oldattr;

	oldattr.hook_enabled = dest->hook_enabled;
	oldattr.hook_enable_left = dest->hook_enable_left;
	oldattr.hook_proc = dest->hook_proc;
	oldattr.handle_method_arity = dest->handle_method_arity;

	dest->hook_enabled = orig->hook_enabled;
	dest->hook_enable_left = orig->hook_enable_left;
	dest->hook_proc = orig->hook_proc;
	dest->handle_method_arity = orig->handle_method_arity;

	if (rb_block_given_p() ) {
		return rb_ensure(rb_yield, Qnil, thread_restore_attributes, (VALUE)&oldattr);
	}

	return Qnil;
}


extern void Init_rallhook_base() {

	const char* initcode = 	"require 'rubygems'\n"
						"require 'ruby-cymbol'\n";

	rb_eval_string(initcode);

/*
This module brings together all classes and methods belonging to rallhook with the exception of Node
*/
	rb_mRallHook = rb_define_module("RallHook");
/*
This class handles the hook, enabling and disable it.

Example:

	# ... instanciate method_handler ... (see README and examples)

	RallHook::Hook.hook method_hadler do
		print "hello world\n" # calls to print, write or whatever are intercepted by method_handler#method_handle
	end # in the finish of the block, the hook are disabled

*/

	rb_cHook = rb_define_class_under(rb_mRallHook, "Hook", rb_cObject);

	rb_define_singleton_method(rb_cHook, "rehook", rehook, 0 );
	rb_define_singleton_method(rb_cHook, "hook", hook, 1);
	rb_define_singleton_method(rb_cHook, "unhook", unhook, 0);
	rb_define_singleton_method(rb_cHook, "from", from, 1);

/*
Marker module to indicate messages of method redirection

Example:

	class MyMessage
		include MethodRedirect # indicates that this message is about method redirection

		def initialize(klass, recv, m)
			@klass = klass
			@recv = recv
			@method = m
		end
	end
	class MethodHandler
		class X
			def foo
			end
		end
		def method_handle( ... )
			MyMessage.new(X, x.new, :foo)
		end
	end

Don't use this module, use x.redirect(:foo) instead.

Example:
	class MethodHandler
		class X
			def foo
			end
		end
		def method_handle( ... )
			x.redirect(:foo)
		end
	end
*/
	rb_mMethodRedirect = rb_define_module_under(rb_mRallHook, "MethodRedirect");

	init_node();
	init_restrict_def();
	init_redirect();

	id_call_ = rb_intern("call");

	id_call = rb_intern("call");
	id_method_wrapper = rb_intern("method_wrapper");
	id_handle_method = rb_intern("handle_method");
	id_return_value_var = rb_intern("@return_value");
	id_klass_var = rb_intern("@klass");
	id_recv_var = rb_intern("@recv");
	id_method_var = rb_intern("@method");
	id_unhook_var = rb_intern("@unhook");
	id_binding = rb_intern("binding");
	id_method_added = rb_intern("method_added");
	id_hook_enabled = rb_intern("__hook_enabled");
	id_hook_enable_left = rb_intern("__hook_enable_left");
	id_hook_proc = rb_intern("__hook_proc");

	rb_define_method(rb_cThread, "acquire_attributes", rb_thread_acquire_attributes,0);

}
